src/Microsoft.AspNet.Mvc.Core/Rendering/Html/TemplateRenderer.cs

```diff
@@ -141,19 +146,31 @@ private IEnumerable<string> GetViewNames()
                 yield return templateHint;
             }
 
+            var fieldType = _viewData.ModelExplorer.ModelType;
+
             // We don't want to search for Nullable<T>, we want to search for T (which should handle both T and
             // Nullable<T>).
-            var modelType = _viewData.ModelExplorer.ModelType;
-            var fieldType = Nullable.GetUnderlyingType(modelType) ?? modelType;
+            fieldType = Nullable.GetUnderlyingType(fieldType) ?? fieldType;
 
-            yield return fieldType.Name;
+            foreach (string typeName in GetTypeNames(_viewData.ModelExplorer, fieldType))
+            {
+                yield return typeName;
+            }
+        }
+
+        public static IEnumerable<string> GetTypeNames(ModelExplorer modelExplorer, Type fieldType)
```

Comment:
pass `ModelMetadata metadata`, not the whole `ModelExplorer`.  only need the metadata.